pkg/traceql/engine_metrics_average.go

```diff
@@ -0,0 +1,540 @@
+// First aggregation layer
+type AvgOverTimeSpanAggregator[F FastStatic, S StaticVals] struct {
+	// Config
+	by              []Attribute   // Original attributes: .foo
+	byLookups       [][]Attribute // Lookups: span.foo resource.foo
+	getSpanAttValue func(s Span) float64
+	start           uint64
+	end             uint64
+	step            uint64
+
+	// Data
+	series     map[F]AvgOverTimeSeries[S]
+	lastSeries AvgOverTimeSeries[S]
+	buf        fastStaticWithValues[F, S]
+	lastBuf    fastStaticWithValues[F, S]
+}
+
+var _ SpanAggregator = (*AvgOverTimeSpanAggregator[FastStatic1, StaticVals1])(nil)
+
+func NewAvgOverTimeSpanAggregator(attr Attribute, by []Attribute, start, end, step uint64) SpanAggregator {
+	lookups := make([][]Attribute, len(by))
+	for i, attr := range by {
+		if attr.Intrinsic == IntrinsicNone && attr.Scope == AttributeScopeNone {
+			// Unscoped attribute. Check span-level, then resource-level.
+			// TODO - Is this taken care of by span.AttributeFor now?
+			lookups[i] = []Attribute{
+				NewScopedAttribute(AttributeScopeSpan, false, attr.Name),
+				NewScopedAttribute(AttributeScopeResource, false, attr.Name),
+			}
+		} else {
+			lookups[i] = []Attribute{attr}
+		}
+	}
+
+	aggNum := len(lookups)
+
+	switch aggNum {
+	case 2:
+		return newAvgAggregator[FastStatic2, StaticVals2](attr, by, lookups, start, end, step)
+	case 3:
+		return newAvgAggregator[FastStatic3, StaticVals3](attr, by, lookups, start, end, step)
+	case 4:
+		return newAvgAggregator[FastStatic4, StaticVals4](attr, by, lookups, start, end, step)
+	case 5:
+		return newAvgAggregator[FastStatic5, StaticVals5](attr, by, lookups, start, end, step)
+	default:
+		return newAvgAggregator[FastStatic1, StaticVals1](attr, by, lookups, start, end, step)
+	}
+}
+
+func newAvgAggregator[F FastStatic, S StaticVals](attr Attribute, by []Attribute, lookups [][]Attribute, start, end, step uint64) SpanAggregator {
+	var fn func(s Span) float64
+
+	switch attr {
+	case IntrinsicDurationAttribute:
+		fn = func(s Span) float64 {
+			return float64(s.DurationNanos()) / float64(time.Second)
+		}
+	default:
+		fn = func(s Span) float64 {
+			f, a := FloatizeAttribute(s, attr)
+			if a == TypeNil {
+				return math.Float64frombits(normalNaN)
+			}
+			return f
+		}
+	}
+
+	return &AvgOverTimeSpanAggregator[F, S]{
+		series:          map[F]AvgOverTimeSeries[S]{},
+		getSpanAttValue: fn,
+		by:              by,
+		byLookups:       lookups,
+		start:           start,
+		end:             end,
+		step:            step,
+	}
+}
+
+func (g *AvgOverTimeSpanAggregator[F, S]) Observe(span Span) {
+	if !g.getGroupingValues(span) {
+		return
+	}
+
+	s := g.getSeries()
+	interval := IntervalOf(span.StartTimeUnixNanos(), g.start, g.end, g.step)
+	if interval == -1 {
+		return
+	}
+	inc := g.getSpanAttValue(span)
+	if math.IsNaN(inc) {
+		return
+	}
+	s.count[interval]++
+	mean, c := averageInc(s.avg[interval], inc, s.count[interval], s.compensation[interval])
```

Comment:
This calculates the mean at span level and bubbles up the count. The other possibility is just sum all the values and calculate the average in the final step

Reply:
I'm leaning towards this approach as well.  Simple summation would be faster but overflow and precision could be worse?  I think we can start with this approach, and switch later, if/when we have objective numbers that the error rate of the other way is better, or this proves to be a bottleneck.